targets/csharp/source/source/PlayFabGameServerInternal.cs

```diff
@@ -0,0 +1,310 @@
+            {
+                Logger.Start();
+            }
+
+            string gsmsBaseUrl = ConfigMap[PlayFabGameServerSDK.HeartbeatEndpointKey];
+            string instanceId = ConfigMap[PlayFabGameServerSDK.ServerIdKey];
+
+            Logger.Log($"VM Agent Endpoint: {gsmsBaseUrl}");
+            Logger.Log($"Instance Id: {instanceId}");
+
+            _webClient = GameServerHttpClientFactory.CreateInstance($"http://{gsmsBaseUrl}/v1/sessionHosts/{instanceId}");
+
+            _heartbeatRunning = true;
+            _signalHeartbeatEvent.Reset();
+            TransitionToActiveEvent.Reset();
+
+            _heartbeatTask = Task.Run(HeartbeatAsync);
+            return Task.CompletedTask;
+        }
+
+        private GameServerConfiguration GetConfiguration()
+        {
+            string fileName = !string.IsNullOrWhiteSpace(_overrideConfigFileName) ? _overrideConfigFileName
+                                : Environment.GetEnvironmentVariable(PlayFabGameServerSDK.GsdkConfigFileEnvVarKey);
+
+            GameServerConfiguration localConfig;
+
+            if (!string.IsNullOrWhiteSpace(fileName) && File.Exists(fileName))
+            {
+                var jsonWrapper = PlayFab.PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer);
+
+                localConfig = new JsonFileConfiguration(jsonWrapper, fileName);
+            }
+            else
+            {
+                localConfig = new EnvironmentVariableConfiguration();
+            }
+
+            return localConfig;
+        }
+
+        private IDictionary<string, string> CreateConfigMap(GameServerConfiguration localConfig)
+        {
+            var finalConfig = new Dictionary<string, string>();
+
+            foreach (KeyValuePair<string, string> certEntry in localConfig.GameCertificates)
+            {
+                finalConfig[certEntry.Key] = certEntry.Value;
+            }
+
+            foreach (KeyValuePair<string, string> metadata in localConfig.BuildMetadata)
+            {
+                finalConfig[metadata.Key] = metadata.Value;
+            }
+
+            foreach (KeyValuePair<string, string> port in localConfig.GamePorts)
+            {
+                finalConfig[port.Key] = port.Value;
+            }
+
+            finalConfig[PlayFabGameServerSDK.HeartbeatEndpointKey] = localConfig.HeartbeatEndpoint;
+            finalConfig[PlayFabGameServerSDK.ServerIdKey] = localConfig.ServerId;
+            finalConfig[PlayFabGameServerSDK.LogFolderKey] = localConfig.LogFolder;
+            finalConfig[PlayFabGameServerSDK.SharedContentFolderKey] = localConfig.SharedContentFolder;
+            finalConfig[PlayFabGameServerSDK.CertificateFolderKey] = localConfig.CertificateFolder;
+            finalConfig[PlayFabGameServerSDK.TitleIdKey] = localConfig.TitleId;
+            finalConfig[PlayFabGameServerSDK.BuildIdKey] = localConfig.BuildId;
+            finalConfig[PlayFabGameServerSDK.RegionKey] = localConfig.Region;
```

Comment:
Is there some sort of defined interface for the finalConfig? It's hard to tell what need's to go from the localConfig into the finalConfig and why. Does everything from the localConfig go into the finalConfig just flattened?

Reply:
that's pretty much it... I wonder if I should include the change to just use a configuration file here

---
In reply to: [256217501](https://github.com/PlayFab/SDKGenerator/pull/398#discussion_r256217501) [](ancestors = 256217501)